src/core_int/integration_coindesk_price.py

```diff
@@ -74,29 +76,50 @@ async def on_contract_created(event):
             })
 
 
-@dataclass
-class IntegrationCoinDeskPriceOracleEnv(IntegrationEnvironment):
-    currencyCode: str
-    updatePeriod: int
+async def update_oracle_commands(cid, currencyCode: str):
+    LOG.debug('update_oracle_commands: %r, %r', cid, currencyCode)
+    resp = await issue_coindesk_request(currencyCode)
 
+    if resp['success']:
+        return [exercise(cid, 'CurrentPriceOracleUpdater_Update', {
+            'updatedAt': resp['updatedAt'],
+            'rate': resp['rate']
+        })]
+    else:
+        LOG.error('Invalid coindesk response (status: %r, body: %r)',
+                  resp['httpStatusCode'], resp['httpResponseBody'])
+        return []
 
 def integration_coindesk_price_oracle_main(
-        env: 'IntegrationCoinDeskPriceOracleEnv',
+        env: 'IntegrationEnvironment',
         events: 'IntegrationEvents'):
 
-    @events.time.periodic_interval(env.updatePeriod)
+    active_oracles = {}
+
+    @events.ledger.contract_created('CoinDesk.PriceOracle:PriceOracleRequest')
+    async def on_contract_created(event):
+        LOG.debug('Noticed oracle request contract: %r (%r)', event.cid, event.cdata)
+
+        currencyCode = event.cdata['currencyCode']
+
+        active_oracles[event.cid] = currencyCode
+
+        return await update_oracle_commands(event.cid, currencyCode)
+
+    @events.ledger.contract_archived('CoinDesk.PriceOracle.PriceOracleRequest')
+    async def on_ledger_archived(event):
+        LOG.debug('Archived oracle request contract: %r', event.cid)
+        active_oracles.pop(event.cid, None)
+
+    @events.time.periodic_interval(30)
     async def poll_coindesk():
-        resp = await issue_coindesk_request(env.currencyCode.upper())
-
-        if not resp['success']:
-            LOG.error('Invalid coindesk response (status: %r, body: %r)',
-                      resp['httpStatusCode'], resp['httpResponseBody'])
-            return []
-
-        return [create_and_exercise(
-            'CoinDesk.PriceOracle.CurrentPriceOracleUpdater',
-            {'integrationParty': env.party},
-            'CurrentPriceOracleUpdater_Update',
-            {'currencyCode': env.currencyCode,
-             'updatedAt': resp['updatedAt'],
-             'rate': resp['rate']})]
+
+        cmds = []
+
+        for (cid, currencyCode) in active_oracles.items():
+            cmds.extend(await update_oracle_commands(cid, currencyCode))
```

Comment:
this is awaiting on each individual `update_oracle_commands` in sequential/blocking fashion, as opposed to awaiting on the result of all of them combined, right?

Reply:
Correct. This is a once-per-minute poll of a slowly updating underlying data source, so I let the simpler straight line code win out over the more parallel approach. I think the most likely issues with this code will be that it doesn't really report poll errors in a proactive way and there are certain sorts of failures that will just break the entire poll cycle. Part of the contract of using this style of integration is to check the recency of the data before you use it.